sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -561,14 +566,18 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
         val origOutput = query.output.head
 
         val resultWithZeroTups = evalSubqueryOnZeroTups(query)
-        if (resultWithZeroTups.isEmpty) {
+        val (topPart, havingNode, aggNode) = splitSubquery(query)
```

Comment:
how about we make the case 1 result a lazy val? (multiple variable lazy val looks weird)
```
lazy val planWithoutCountBug = Project(... // Or just val as constructing logical plan is cheap
if (resultWithZeroTups.isEmpty) {
  planWithoutCountBug
} else {
  val (topPart, havingNode, aggNode) = splitSubquery(query)
  if (aggNode.isEmpty) planWithoutCountBug else ...
}

```